examples/image_ocr.py

```diff
@@ -0,0 +1,433 @@
+            ret.append(ord(char) - ord('a'))
+        elif char == ' ':
+            ret.append(26)
+    return ret
+
+# only a-z and space..probably not to difficult
+# to expand to uppercase and symbols
+def is_valid_str(in_str):
+    search = re.compile(r'[^a-z\ ]').search
+    return not bool(search(in_str))
+
+# Uses generator functions to supply train/test with
+# data. Image renderings are text are created on the fly
+# each time with random perturbations 
+class TextImageGenerator(keras.callbacks.Callback):
+    ABSOLUTE_MAX_STRING_LEN = 16
+
+    def __init__(self, monogram_file, bigram_file, minibatch_size, img_w,
+                 img_h, downsample_width,  num_words, val_split):
+        assert(num_words % minibatch_size == 0)
+        assert((val_split * num_words) % minibatch_size == 0)
+
+        self.minibatch_size = minibatch_size
+        self.img_w = img_w
+        self.img_h = img_h
+        self.monogram_file = monogram_file
+        self.bigram_file = bigram_file
+        self.downsample_width = downsample_width
+        self.num_words = num_words
+        self.val_split = val_split
+        self.blank_label = self.get_output_size()-1
+
+    def get_output_size(self):
+        return 28
+
+    def build_word_list(self, max_string_len=None, mono_fraction=0.5):
+        assert(max_string_len <= TextImageGenerator.ABSOLUTE_MAX_STRING_LEN)
+        self.string_list = []
+        self.max_string_len = max_string_len
+        self.Y_data = np.ones([self.num_words, TextImageGenerator.ABSOLUTE_MAX_STRING_LEN]) * -1
+        self.X_text = []
+        self.Y_len = [0] * self.num_words
+
+        # monogram file is sorted so just read first N words
+        # since some later words are nonsensical
+        with open(self.monogram_file, "rt") as f:
+            for line in f:
+                if(len(self.string_list) == int(self.num_words * mono_fraction)):
+                    break
+                word = line.rstrip()
+                if(max_string_len == -1 or max_string_len is None or len(word) <= max_string_len):
+                    self.string_list.append(word)
+        # bigram file doesn't contain nonsensical words,
+        # so read the entire thing and pull some random lines out
+        with open(self.bigram_file, "rt") as f:
+            lines = f.readlines()
+            random.shuffle(lines)
+            for line in lines:
+                if(len(self.string_list) == self.num_words):
+                    break
+                columns = line.lower().split()
+                word = columns[0] + " " + columns[1]
+                if(is_valid_str(word) and
+                        (max_string_len == -1 or max_string_len is None or len(word) <= max_string_len)):
+                    self.string_list.append(word)
+        if (len(self.string_list) != self.num_words):
+            raise IOError("Could not pull enough words from supplied monogram and bigram files. ")
+        random.shuffle(self.string_list)
```

Comment:
This is true, though the font, translation, rotation, speckle, etc. would be different.  Nevertheless I did notice that the RNNs learn some sort of language model since results are worse with random characters.  So this is probably worth fixing.